### ISSUE TITLE
Incorrect regex label match when using `|` with an empty option

### ISSUE THREAD

**Issue by risinger** [thanos-io/thanos#4646]
**Thanos, Prometheus and Golang version used**:
Thanos: v0.22.0
Prometheus: v2.28.1

**Object Storage Provider**:
S3

**What happened**:
Using a Thanos Store Gateway as the data source, `prometheus_build_info{doesnotexist=~"whocares|"}` (note the tailing `|` on that regex) returns no series. 

**What you expected to happen**:
I expected the Store Gateway to return the same series as `prometheus_build_info{doesnotexist=""}`, which is what a Prometheus data source does. 

**How to reproduce it (as minimally and precisely as possible)**:
Write a query with `label=~"value|"` on a metric without a series with `label="value"`.
Use Store Filtering in the Query UI to select either a Prometheus instance or a Store Gateway. 
Example query:
`prometheus_build_info{doesnotexist=~"whocares|"}`

**Full logs to relevant components**:
Not applicable, I don't think. 

**Anything else we need to know**:
Basically Prometheus appears to interpret `{label=~"value|"}` as either `{label=value}` or `{label=""}` (label does not exist). I expected Thanos to do the same. 

Let me know if you need anymore info.

**Comment by GiedriusS** [comment]
I was able to reproduce this. Interesting bug. Thanks for the report! Help welcome.

**Comment by yeya24** [comment]
The reason is that our optimization on regex is not exactly the same as Prometheus' one.

Our implementation: https://github.com/thanos-io/thanos/blob/4276b50564d745d28a9f6ce3e92d19d6fd50c46f/pkg/store/bucket.go#L1763.

Prometheus: https://github.com/prometheus/prometheus/blob/b1ed4a0a663d0c62526312311c7529471abbc565/tsdb/querier.go#L221

As `{doesnotexist=~"whocares|"}` matches the empty string `""`, Prometheus will go for an inverse posting match. However, Thanos doesn't check whether the empty string is matched or not so the logic is different.

**Comment by kunaljain0212** [comment]
I would like to work on this issue, could you please guide me about how should I proceed.

**Comment by GiedriusS** [comment]
I suppose a good start would be to add failing tests. A suitable location seems to be the e2e tests: https://github.com/thanos-io/thanos/tree/main/test/e2e.

**Comment by Himanshu2107** [comment]
I would like to take up this issue.

**Comment by drstevens** [comment]
@Himanshu2107 are you working on this?

**Comment by Himanshu2107** [comment]
@drstevens no, not right now.

**Comment by drstevens** [comment]
Bump

**Comment by MichaHoffmann** [comment]
I think this might be fixed by #6692 !

**Comment by drstevens** [comment]
Thanks @MichaHoffmann 🍻

**Comment by MichaHoffmann** [comment]
> Thanks @MichaHoffmann 🍻

:beers:

**Comment by GiedriusS** [comment]
🏅 @MichaHoffmann
